c2rust-transpile/src/translator/mod.rs

```diff
@@ -2231,6 +2231,22 @@ impl<'c> Translation<'c> {
                     .unwrap_or((*id, ty));
                 let expr = self.convert_expr(ctx, expr_id)?;
 
+                // Issue: https://github.com/immunant/c2rust/issues/803
+                // This actually should be catched by `to_unsafe_pure_expr` see below
+                // but a function call expression is translated as having no sideeffects by `convert_expr` (above)
+                // which is not necessarily correct, we just might not know them (e.g only decl in header exists).
+                // So we catch this here until there is a way to express unknown sideeffects of an expression.
```

Comment:
The purpose of `to_unsafe_pure_expr` is not actually to enforce the kind of purity that Rust `const`s require. It checks a notion of purity based on "does the expression computing this value have prerequisite statements that we need to ensure are evaluated prior to the expression itself".

We sadly don't have a good notion of "is this Rust expression const-safe", and it's difficult to build one because `c2rust` is oriented toward reasoning about C code, not Rust code. See [this comment](https://github.com/immunant/c2rust/issues/803#issuecomment-1544754766).